### ISSUE TITLE
Idea: Setup routes to remove the vm ip annoyance

### ISSUE THREAD

**Issue by carlsverre** [nlf/dlite#28]
On my personal docker setup I use something similar to the following command to directly expose my docker container's ip's to my mac.  It has made playing with docker on mac much similar to how it feels on linux when you can pretty much treat a container as a lightweight vm.

```
sudo route -n add DOCKER_SUBNET DOCKER_VM_IP
```

The subnet is usually something like 172.0.0.0/16 by default but check first.

The route is removed when you restart and can be removed by a mirror command like so:

```
sudo route -n delete DOCKER_SUBNET
```

I love this tool you have put together, and I think this feature would get it that much closer to "just using" docker on a mac.  :)

**Comment by arnested** [comment]
@carlsverre, I'm trying to setup routes myself but I cannot get the vm host to forward the packets to the container.

It worked for me flawlessly on my docker-machine/virtualbox-setup so any advice on how you achieve this would be appreciated.

@nlf, dlite looks really nice!

**Comment by carlsverre** [comment]
Whoops - totally forgot that I have forwarding rules setup on my docker vm.  I am pretty sure that either docker-machine or the docker daemon does this.  Anyways, here is what works for my docker-machine vm: 

```
docker@dmachine:~$ sudo iptables -S
-P INPUT ACCEPT
-P FORWARD ACCEPT
-P OUTPUT ACCEPT
-N DOCKER
-A FORWARD -o docker0 -j DOCKER
-A FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i docker0 ! -o docker0 -j ACCEPT
-A FORWARD -i docker0 -o docker0 -j ACCEPT
```

**Comment by arnested** [comment]
I have the same on my dlite vm and haven't been able to spot any other differences from my docker-machine vm either. But no working packet routing ...

I'll keep on digging ... something must be different!

**Comment by nlf** [comment]
this is an interesting idea for sure. let me know if you figure out how to get the routes working! also it may be of use to go through old issues in the [xhyve](https://github.com/mist64/xhyve) repo where i feel like this has been brought up before

**Comment by arnested** [comment]
Bingo, @nlf!

Your advice let me to ailispaw/boot2docker-xhyve/issues/7 and that works for me as well.

```
$ sudo route -n add 172.17.0.0/16 local.docker
add net 172.17.0.0: gateway local.docker
$ ifconfig bridge100
bridge100: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
    options=3<RXCSUM,TXCSUM>
    ether ba:e8:56:63:23:64
    inet 192.168.64.1 netmask 0xffffff00 broadcast 192.168.64.255
    Configuration:
        id 0:0:0:0:0:0 priority 0 hellotime 0 fwddelay 0
        maxage 0 holdcnt 0 proto stp maxaddr 100 timeout 1200
        root id 0:0:0:0:0:0 priority 0 ifcost 0 port 0
        ipfilter disabled flags 0x2
    member: en12 flags=3<LEARNING,DISCOVER>
            ifmaxaddr 0 port 17 priority 0 path cost 0
    Address cache:
        a6:84:b2:34:cf:32 Vlan1 en12 1170 flags=0<>
    nd6 options=1<PERFORMNUD>
    media: autoselect
    status: active
$ sudo ifconfig bridge100 -hostfilter en12
$ ping 172.17.0.2
PING 172.17.0.2 (172.17.0.2): 56 data bytes
64 bytes from 172.17.0.2: icmp_seq=0 ttl=63 time=0.513 ms
64 bytes from 172.17.0.2: icmp_seq=1 ttl=63 time=0.324 ms
^C
--- 172.17.0.2 ping statistics ---
2 packets transmitted, 2 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 0.324/0.418/0.513/0.095 ms
```

**Comment by arnested** [comment]
And the hostfilter part as a shell one-liner (although not pretty):

``` sh
INTERFACE=$(route get local.docker | grep interface: | cut -f 2 -d:); sudo ifconfig ${INTERFACE} -hostfilter $(ifconfig ${INTERFACE} | grep member: | cut -f 2 -d: | cut -f 2 -d\ )
```

**Comment by carlsverre** [comment]
Wow good find! Would be sweet if xhyve was able to set that up on the bridge automatically. But for now the one liner works perfect. Maybe we should add this to a faq somewhere?

**Comment by ruimarinho** [comment]
Nice one @arnested! :) I'm also a converted docker-machine user and routing was half of the experience. What would be the best way of automating this step?

**Comment by chris-codaio** [comment]
+1 for enabling this!

**Comment by nlf** [comment]
It looks like it's safe to add the route and set the hostfilter repeatedly, so I'm gonna go ahead and implement this (and the dnsdock feature) within DLite automatically, but behind a configuration option so it's opt-in and runs when the daemon is launched.

**Comment by ruimarinho** [comment]
I don't know how often routes change on OSX, but I do know that they are reset upon certain events (WiFi connect/disconnect? VPNs?). Since I believe monitoring by DLite is outside its scope, perhaps a "route" command could be used to refresh routes?

**Comment by nlf** [comment]
yeah, that might be necessary

**Comment by nlf** [comment]
looking at your modified instructions that loop and add the hostfilter for every interface that's a member of the bridge interface, are there any negative side effects of adding the hostfilter for the incorrect interface?

if so, for simplicity i'm thinking i just set the hostfilter for every member of the bridge and call it a day rather than attempting the pings

**Comment by nlf** [comment]
closed via 94751cc5c6807fdce70de33864d38799d9d74615
